File: tests/ut/cpp/python_input/gtest_input/side_effect/order_enforce_test.py
# Copyright 2022 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
""" order_enforce_test """
from mindspore import Tensor
from mindspore import dtype as mstype
from mindspore.ops import operations as P
from mindspore.ops import functional as F
from mindspore import Parameter


# pylint: disable=unused-variable

class TestOrderEnforceFnDict:
    def __init__(self):
        self.fn_dict = {}

    def __call__(self, fn):
        self.fn_dict[fn.__name__] = fn

    def __getitem__(self, name):
        try:
            return self.fn_dict[name]
        except KeyError:
            return None


# Test ut for file: order_enforce.cc.
def test_order_enforce(tag):
    """
    Feature: Side Effect.
    Description: Insert `TensorMove` nodes afters some special `Load` nodes.
    Expectation: The `TensorMove`nodes are inserted after `Load` nodes as expected.
    """
    fns = TestOrderEnforceFnDict()

    param_a = Parameter(Tensor([1], mstype.int32), name='a')
    param_b = Parameter(Tensor([2], mstype.int32), name='b')
    param_c = Parameter(Tensor([3], mstype.int32), name='c')
    param_d = Parameter(Tensor([4], mstype.int32), name='d')

    x = Tensor([5], mstype.int32)
    y = Tensor([6], mstype.int32)

    load = P.Load()
    pow_op = P.Pow()

    @fns
    def test_two_loads():
        load1 = load(param_a)
        load2 = load(param_a)
        load3 = load(param_b)
        return load1, load2, load3

    @fns
    def test_partial_load_arg():
        def partial_func(parameter_0, parameter_1):
            return parameter_0, parameter_1

        func = F.partial(partial_func, load(param_a))
        output = func(load(param_b))
        return output

    @fns
    def test_partial_load_arg_call_out_as_arg():
        def partial_func(parameter_0, parameter_1):
            return parameter_0, parameter_1

        def arg_func():
            return load(param_a)

        func = F.partial(partial_func, arg_func())
        output = func(load(param_b))
        return output

    @fns
    def test_call_out_load():
        def partial_func(parameter_0, parameter_1):
            return parameter_0, parameter_1

        func = F.partial(partial_func, param_a)
        output = func(param_b)
        return load(F.tuple_getitem(output, 0)), load(F.tuple_getitem(output, 1))

    @fns
    def load_ref_same_to_call_out():
        def true_func():
            return (param_a, param_d), (param_c, param_c)

        def false_func():
            return (param_b, param_d), (param_c, param_c)

        cond = pow_op(x, x) < pow_op(y, y)
        func = F.switch(cond, true_func, false_func)
        switch_call_output = func()
        tuple_getitem_0_0 = F.tuple_getitem(F.tuple_getitem(switch_call_output, 0), 0)
        tuple_getitem_0_1 = F.tuple_getitem(F.tuple_getitem(switch_call_output, 0), 1)
        output = F.make_tuple(load(param_a), load(param_b), load(param_c), load(tuple_getitem_0_0),
                              load(tuple_getitem_0_1))
        return output

    @fns
    def test_switch_switch_call():
        def true_func():
            def true_true_func():
                return param_a

            def true_false_func():
                return param_b

            cond = pow_op(x, x) < pow_op(y, y)
            return F.switch(cond, true_true_func, true_false_func)()

        def false_func():
            def false_true_func():
                return param_c

            def false_false_func():
                return param_c

            cond = pow_op(x, x) < pow_op(y, y)
            return F.switch(cond, false_true_func, false_false_func)()

        def func():
            cond = pow_op(x, x) < pow_op(y, y)
            return F.switch(cond, true_func, false_func)()

        output = F.make_tuple(load(param_a), load(param_b), load(param_c), load(param_d), load(func()))
        return output

    return fns[tag]

File: tests/ut/cpp/side_effect/order_enforce_test.cc
/**
 * Copyright 2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <memory>

#include "pipeline/jit/ps/static_analysis/order_enforce.h"

#include "mindspore/core/ops/framework_ops.h"
#include "common/common_test.h"
#include "common/py_func_graph_fetcher.h"

#include "ir/anf.h"
#include "ir/visitor.h"
#include "pipeline/jit/ps/action.h"
#include "ir/func_graph.h"
#include "include/common/debug/anf_ir_dump.h"

namespace mindspore {
namespace opt {
class TestSideEffectOrderEnforce : public UT::Common {
 public:
  TestSideEffectOrderEnforce() : getPyFun("gtest_input.side_effect.order_enforce_test", true) {}

  void SetUp() {}

 public:
  UT::PyFuncGraphFetcher getPyFun;
};

void InsertUNode(const FuncGraphPtr &func_graph) {
  auto nodes = TopoSort(func_graph->get_return(), SuccDeeper);
  for (const auto &node : nodes) {
    if (IsPrimitiveCNode(node, prim::kPrimLoad)) {
      auto load = node->cast<CNodePtr>();
      load->add_input(NewValueNode(kUMonad));
    }
  }
}

size_t CountTensorMoveNum(const FuncGraphPtr &func_graph) {
  auto nodes = TopoSort(func_graph->get_return(), SuccDeeperSimple);
  auto accumulate_func = [](size_t prev_num, const AnfNodePtr &node) -> size_t {
    return IsPrimitiveCNode(node, prim::kPrimTensorMove) ? prev_num + 1 : prev_num;
  };
  return std::accumulate(nodes.begin(), nodes.end(), 0, accumulate_func);
}

std::vector<size_t> GetInsertTensorMoveIndexes(const FuncGraphPtr &func_graph) {
  HashMap<AnfNodePtr, size_t> load_index;
  std::vector<size_t> indexes;
  size_t cur_index = 0;
  auto nodes = TopoSort(func_graph->get_return(), SuccDeeperSimple);
  for (const auto &node : nodes) {
    if (IsPrimitiveCNode(node, prim::kPrimLoad)) {
      load_index[node] = cur_index++;
      continue;
    }
    if (IsPrimitiveCNode(node, prim::kPrimTensorMove)) {
      auto tensor_move = node->cast<CNodePtr>();
      auto load = tensor_move->input(1);
      auto it = load_index.find(load);
      if (it == load_index.end()) {
        MS_LOG(EXCEPTION) << "Can't find load as input of tensor move.";
      }
      indexes.push_back(it->second);
    }
  }
  return indexes;
}

// Feature: Insert tensor move.
// Description: If there are 2 or more `Load` nodes with same ref key, `TensorMove` nodes need inserted after these
// `Load` nodes.
// Expectation: `TensorMove` nodes inserted after `Load` nodes with same ref key of which nums greater than 2.
TEST_F(TestSideEffectOrderEnforce, TwoLoads) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_order_enforce", "test_two_loads");
  ASSERT_TRUE(nullptr != test_graph);

  FuncGraphManagerPtr manager1 = Manage(test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  res->set_func_graph(test_graph);
  std::vector<AbstractBasePtr> args_spec;

  // Execute pass
  InsertUNode(test_graph);
  pipeline::AbstractSpecializeAction(res);
  pipeline::OrderEnforce(res->func_graph());

  // Check tensor move indexes.
  auto tensor_move_indexes = GetInsertTensorMoveIndexes(res->func_graph());
  std::vector<size_t> expect_indexes = {0, 1, 2};
  ASSERT_EQ(tensor_move_indexes.size(), expect_indexes.size());
  auto result = tensor_move_indexes == expect_indexes;
  ASSERT_TRUE(result);

  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}

// Feature: Insert tensor move.
// Description: If the arg of `Partial` or func graph call node is `Load`, insert a `TensorMove` node after it.
// Expectation: `TensorMove` nodes inserted after `Load` nodes of which are arg of `Partial` or func graph call node.
TEST_F(TestSideEffectOrderEnforce, PartialArgCallArg) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_order_enforce", "test_partial_load_arg");
  ASSERT_TRUE(nullptr != test_graph);

  FuncGraphManagerPtr manager1 = Manage(test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  res->set_func_graph(test_graph);
  std::vector<AbstractBasePtr> args_spec;

  // Execute pass
  InsertUNode(test_graph);
  pipeline::AbstractSpecializeAction(res);
  pipeline::OrderEnforce(res->func_graph());

  // Check tensor move indexes.
  auto tensor_move_indexes = GetInsertTensorMoveIndexes(res->func_graph());
  std::vector<size_t> expect_indexes = {0, 1};
  ASSERT_EQ(tensor_move_indexes.size(), expect_indexes.size());
  auto result = tensor_move_indexes == expect_indexes;
  ASSERT_TRUE(result);

  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}

// Feature: Insert tensor move.
// Description: If the arg of `Partial` or func graph call node is `Load`, insert a `TensorMove` node after it.In this
// case, an arg is output of func graoh call.
// Expectation: `TensorMove` nodes inserted after `Load` nodes of which are arg of `Partial` or func graph call node.
TEST_F(TestSideEffectOrderEnforce, CallOutAsArg) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_order_enforce", "test_partial_load_arg_call_out_as_arg");
  ASSERT_TRUE(nullptr != test_graph);

  FuncGraphManagerPtr manager1 = Manage(test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  res->set_func_graph(test_graph);
  std::vector<AbstractBasePtr> args_spec;

  // Execute pass
  InsertUNode(test_graph);
  pipeline::AbstractSpecializeAction(res);
  pipeline::OrderEnforce(res->func_graph());

  // Check tensor move indexes.
  auto tensor_move_indexes = GetInsertTensorMoveIndexes(res->func_graph());
  std::vector<size_t> expect_indexes = {0, 1};
  ASSERT_EQ(tensor_move_indexes.size(), expect_indexes.size());
  auto result = tensor_move_indexes == expect_indexes;
  ASSERT_TRUE(result);

  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}

// Feature: Insert tensor move.
// Description: A func graph call node output is abstract ref and as a input of load.
// Expectation: `TensorMove` nodes inserted after these `Load` nodes.
TEST_F(TestSideEffectOrderEnforce, DISABLED_CallOutLoad) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_order_enforce", "test_call_out_load");
  ASSERT_TRUE(nullptr != test_graph);

  FuncGraphManagerPtr manager1 = Manage(test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  res->set_func_graph(test_graph);
  std::vector<AbstractBasePtr> args_spec;

  // Execute pass
  InsertUNode(test_graph);
  pipeline::AbstractSpecializeAction(res);
  pipeline::OrderEnforce(res->func_graph());

  // Check tensor move indexes.
  auto tensor_move_indexes = GetInsertTensorMoveIndexes(res->func_graph());
  std::vector<size_t> expect_indexes = {0, 1};
  ASSERT_EQ(tensor_move_indexes.size(), expect_indexes.size());
  auto result = tensor_move_indexes == expect_indexes;
  ASSERT_TRUE(result);

  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}

// Feature: Insert tensor move.
// Description: A func graph call node output is abstract ref and as a input of load and has a same refkey with another
// simple `Load` node.
// Expectation: `TensorMove` nodes inserted after these `Load` nodes whose input is a func graph call node and after the
// same refkey simple loads.
TEST_F(TestSideEffectOrderEnforce, DISABLED_LoadRefSameToSwitchCallOutGetItemGetItem) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_order_enforce", "load_ref_same_to_call_out");
  ASSERT_TRUE(nullptr != test_graph);

  FuncGraphManagerPtr manager1 = Manage(test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  res->set_func_graph(test_graph);
  std::vector<AbstractBasePtr> args_spec;

  // Execute pass
  InsertUNode(test_graph);
  pipeline::AbstractSpecializeAction(res);
  pipeline::OrderEnforce(res->func_graph());

  // Check tensor move indexes.
  auto tensor_move_indexes = GetInsertTensorMoveIndexes(res->func_graph());
  std::vector<size_t> expect_indexes = {0, 1, 3, 4};
  ASSERT_EQ(tensor_move_indexes.size(), expect_indexes.size());
  auto result = tensor_move_indexes == expect_indexes;
  ASSERT_TRUE(result);

  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}

// Feature: Insert tensor move.
// Description: A func graph call node output is abstract ref and as a input of load and has a same refkey with another
// simple `Load` node.In this case, call node is a switch_switch call node.
// Expectation: `TensorMove` nodes inserted after these `Load` nodes whose input is a func graph call node and after the
// same refkey simple loads.
TEST_F(TestSideEffectOrderEnforce, DISABLED_LoadCallSwitchCallSwitchCallOut) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_order_enforce", "test_switch_switch_call");
  ASSERT_TRUE(nullptr != test_graph);

  FuncGraphManagerPtr manager1 = Manage(test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  res->set_func_graph(test_graph);
  std::vector<AbstractBasePtr> args_spec;

  // Execute pass
  InsertUNode(test_graph);
  pipeline::AbstractSpecializeAction(res);
  pipeline::OrderEnforce(res->func_graph());
  // Check tensor move indexes.
  auto tensor_move_indexes = GetInsertTensorMoveIndexes(res->func_graph());
  std::vector<size_t> expect_indexes = {0, 1, 2, 4};
  ASSERT_EQ(tensor_move_indexes.size(), expect_indexes.size());
  auto result = tensor_move_indexes == expect_indexes;
  ASSERT_TRUE(result);

  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}
}  // namespace opt
}  // namespace mindspore
